sympy/sets/tests/test_sets.py

```diff
@@ -1597,6 +1600,42 @@ def test_DisjointUnion_len():
     assert len(DisjointUnion(S.EmptySet, S.EmptySet, FiniteSet(x, y, z), S.EmptySet)) == 3
     raises(ValueError, lambda: len(DisjointUnion(Interval(0, 1), S.EmptySet)))
 
+def test_SetKind_ProductSet():
+    p = ProductSet(FiniteSet(Matrix([1, 2])), FiniteSet(Matrix([1, 2])))
+    mk = MatrixKind(NumberKind)
+    k = SetKind(TupleKind(mk, mk))
+    assert p.kind is k
+    assert ProductSet(Interval(1, 2), FiniteSet(Matrix([1, 2]))).kind is SetKind(TupleKind(NumberKind, mk))
+
+def test_SetKind_Interval():
+    assert Interval(1, 2).kind is SetKind(NumberKind)
+
+def test_SetKind_EmptySet_UniversalSet():
+    assert S.UniversalSet.kind is SetKind(UndefinedKind)
+    assert EmptySet.kind is SetKind()
+
+def test_SetKind_FiniteSet():
+    assert FiniteSet(1, Matrix([1, 2])).kind is SetKind(UndefinedKind)
+    assert FiniteSet(1, 2).kind is SetKind(NumberKind)
+
+def test_SetKind_Unions():
+    assert Union(FiniteSet(Matrix([1, 2])), Interval(1, 2)).kind is SetKind(UndefinedKind)
+    assert Union(Interval(1, 2), Interval(1, 7)).kind is SetKind(NumberKind)
+
+def test_SetKind_evaluate_False():
+    one = Union({1}, EmptySet, evaluate=False).kind
+    two = Union(Interval(1, 2), EmptySet, evaluate=False).kind
+    three = Intersection({1}, S.UniversalSet,evaluate=False).kind
+    four = Union({1}, S.UniversalSet, evaluate=False).kind
+    five = Union(Interval(1, 2), Interval(4, 5), FiniteSet(1), evaluate=False).kind
+    six = Intersection({1}, EmptySet, evaluate=False).kind
+    assert one is SetKind(NumberKind)
+    assert two is SetKind(NumberKind)
+    assert three is SetKind(NumberKind)
+    assert four is SetKind(UndefinedKind)
+    assert five is SetKind(NumberKind)
+    assert six is SetKind()
```

Comment:
I find this way of writing tests hard to read because you have made the output far away from the input e.g. you define `one` at the top but then don't use it until 6 lines down after lots of irrelevant things. It would be better to group together the expected output with the test of that output.